### ISSUE TITLE
Divergence between source/build/package methods

### ISSUE THREAD

**Issue by abuszta** [conan-io/conan#604]
Currently you can test portions of your conanfile.py using following functions:
- source downloading: conan source < package_reference >
- package building: conan build < package_directory >
- packaging: conan package < package_reference >

It is very useful to have all of them, but currently conan source/package require you to have package exported but conan build does not.

Would be possible to allow conan source/package be available also for local usage (without having package exported)? It would allow testing recipes before exporting them into the system. E.g.:

```
cd my_package
mkdir developing_recipe && cd developing_recipe
conan source ..
... # May fix something
conan build ..
... # May fix something
conan package ..
... # May fix something
cd ..
conan export abuszta/stable # Ok! I know everything works, I can move to other packages
```

I know that I can use test_package, but this command is doing all that steps combined which takes away possibility of this iterative way of working which can be very useful when working with huge amount of dependencies.

**Comment by memsharded** [comment]
The problem is the scope and the folder layout. In the local cache, they are different folders: the export, the source, the build and the package folders.

When you work in your user space, there is no package folder at all, where to copy the artifacts. Conan build can be executed because it assumes that you are currently in the build folder (like a local build folder to build out-of-source). I am not sure how the source command should behave in the user spaces.

The difference is that such ``build`` command is not intended for package creators, but for package consumers, it is just an automation step beyond the simple ``conanfile.txt`` and manually invoking ``cmake`` or other build system. While the other commands are intended for package creators while debugging and testing. Actually, the ``package`` command only makes sense if it is a real package, that can be consumed for use (i.e. exported in the local conan cache), otherwise it is impossible to test that the package is actually working fine.

I will have another look, but this would be the rationale behind, please tell me if it makes sense. Thanks!

**Comment by abuszta** [comment]
I see... conan package totally does not fit the use case. It is transferring stuff from build directory to export directory - understood. Therefore it makes no sense using it with local directory.

However I'm not so sure about source. My use case is that I was trying to switch for my packages from exports = '*' solution to source method. I'm not keeping recipe with sources, because I'm not the owner of sources (in this case). I'm additionally using VCS to keep track of recipes which I created.

So when I would like to work with some package as a consumer (contribute to project itself, try to fix some issue) I need to:
- download recipe from conan server (or from VCS - it may be not exported/uploaded yet)
- download manually package sources from VCS (and this may be more complex than just simple "git clone")
- run conan install (download dependencies)
- run conan build

What I would like to do is:
- download recipe from conan server or VCS
- run conan source
- run conan install (download dependencies)
- run conan build

In this way I have unified interface for working with all available packages. I don't care if this package source has to be obtained from SVN, CVS, git or using tools.download. Work can be as simple as running conan source/install/build.

> The difference is that such build command is not intended for package creators, but for package consumers, it is just an automation step beyond the simple conanfile.txt and manually invoking cmake or other build system.

From my perspective there should be very little difference between package creator and consumer. Currently in C++ world working with bunch of different projects is still **hard**. People are using different VCS, different build systems, different testing tools. Conan makes that stuff **easy** when you are using some projects as your dependency (just write requires = 'package/0.1@abuszta/testing' and that's all). Working with package/dependency as project (e.g. when you want to contribute) is currently not as straight forward. Having clean and easy interface for working with packages as projects will enable people willingly share their work with others. Sounds feasible?

**Comment by memsharded** [comment]
Hi Antoni,

I think that the divergence was quite addressed in the latest PR #624, it even includes some tests that show how to do the source/build/package completely in user space, with some other fixes too. They will be included in next 0.15 release, but if you want to try them right now, try running conan from source, "develop" branch. I will mark the issue as fixed (we close them when released), please tell me any further feedback or request, thanks!

**Comment by abuszta** [comment]
Yes, I saw this PR and it looks good to me! I will get back with any additional feedback if needed. Thanks :)
